packages/vote/fpc/fpc.go

```diff
@@ -0,0 +1,267 @@
+	ErrNoOpinionGiversAvailable = errors.New("can't perform round as no opinion givers are available")
+)
+
+// New creates a new FPC instance.
+func New(opinionGiverFunc vote.OpinionGiverFunc, paras ...*Parameters) *FPC {
+	f := &FPC{
+		opinionGiverFunc: opinionGiverFunc,
+		paras:            DefaultParameters(),
+		rng:              rand.New(rand.NewSource(time.Now().UnixNano())),
+		ctxs:             make(map[string]*VoteContext),
+		queue:            list.New(),
+		queueSet:         make(map[string]struct{}),
+		events: vote.Events{
+			Finalized: events.NewEvent(vote.OpinionCaller),
+			Failed:    events.NewEvent(vote.OpinionCaller),
+			Error:     events.NewEvent(events.ErrorCaller),
+		},
+	}
+	if len(paras) > 0 {
+		f.paras = paras[0]
+	}
+	return f
+}
+
+// FPC is a DRNGRoundBasedVoter which uses the Opinion of other entities
+// in order to finalize an Opinion.
+type FPC struct {
+	events           vote.Events
+	opinionGiverFunc vote.OpinionGiverFunc
+	// the lifo queue of newly enqueued items to vote on.
+	queue *list.List
+	// contains a set of currently queued items.
+	queueSet map[string]struct{}
+	queueMu  sync.Mutex
+	// contains the set of current vote contexts.
+	ctxs   map[string]*VoteContext
+	ctxsMu sync.RWMutex
+	// parameters to use within FPC.
+	paras *Parameters
+	// indicates whether the last round was performed successfully
+	lastRoundCompletedSuccessfully bool
+	rng                            *rand.Rand
+}
+
+func (f *FPC) Vote(id string, initOpn vote.Opinion) error {
+	f.queueMu.Lock()
+	defer f.queueMu.Unlock()
+	f.ctxsMu.RLock()
+	defer f.ctxsMu.RUnlock()
+	if _, alreadyQueued := f.queueSet[id]; alreadyQueued {
+		return fmt.Errorf("%w: %s", ErrVoteAlreadyOngoing, id)
+	}
+	if _, alreadyOngoing := f.ctxs[id]; alreadyOngoing {
+		return fmt.Errorf("%w: %s", ErrVoteAlreadyOngoing, id)
+	}
+	f.queue.PushBack(newVoteContext(id, initOpn))
+	f.queueSet[id] = struct{}{}
+	return nil
+}
+
+func (f *FPC) IntermediateOpinion(id string) (vote.Opinion, error) {
+	f.ctxsMu.RLock()
+	defer f.ctxsMu.RUnlock()
+	voteCtx, has := f.ctxs[id]
+	if !has {
+		return vote.Unknown, fmt.Errorf("%w: %s", vote.ErrVotingNotFound, id)
+	}
+	return voteCtx.LastOpinion(), nil
+}
+
+func (f *FPC) Events() vote.Events {
+	return f.events
+}
+
+// Round enqueues new items, sets opinions on active vote contexts, finalizes them and then
+// queries for opinions.
+func (f *FPC) Round(rand float64) error {
+	// enqueue new voting contexts
+	f.enqueue()
+	// we can only form opinions when the last round was actually executed successfully
+	if f.lastRoundCompletedSuccessfully {
+		// form opinions by using the random number supplied for this new round
+		f.formOpinions(rand)
+		// clean opinions on vote contexts where an opinion was reached in FinalizationThreshold
+		// number of rounds and clear those who failed to be finalized in MaxRoundsPerVoteContext.
+		f.finalizeOpinions()
+	}
+	// query for opinions on the current vote contexts
```

Comment:
Here we could trigger an event EndRound and include the following info:
* the `rand` used
* the list of `IDs` object of voting
* the list of `opinionGivers` with their opinions
* the formed opinion for all of the `IDs`

Reply:
I've added a `RoundExecuted` event which gives you the rand used, the active vote contexts and the opinion givers with their corresponding response opinion. Additionally added a duration field and how many times a given opinion giver's opinion was taken into account.